### PR TITLE
Support TLS in TCP connections

### PR DESCRIPTION
**Pull request by hudashot**
Also, report certificate expiration date just like HTTPS probe does.

### CODE REVIEW

File: tcp.go [hunk 1]
```diff
@@ -11,19 +12,34 @@ import (
 	"github.com/prometheus/log"
 )
 
+func dialTCP(target string, ssl bool, timeout time.Duration) (net.Conn, error) {
+	dialer := &net.Dialer{Timeout: timeout}
+	if !ssl {
+		return dialer.Dial("tcp", target)
+	}
+	config := &tls.Config{InsecureSkipVerify: true}
```
Comment (brian-brazil) [inline tcp.go]: This needs to be secure by default.

Similar to #23, if we're going to be offering non-default SSL options they should be common across prometheus/blackbox_exporter/alertmanger.

Comment (hudashot) [inline tcp.go]: Not sure I understand what you mean by "common".

There are some comments in #23 about passing options from prometheus to blackbox_exporter using GET parameters to `/probe`. Is this how you suggest to set InsecureSkipVerify?

Comment (brian-brazil) [inline tcp.go]: It should all be in the config file.

There's no mention of GET in that PR, the only things that should be passed in are module name and target.

**Comment by hudashot** [comment]
@brian-brazil, PTAL.

I think if `TLSConfig` in prometheus is supposed to be re-usable in other components it might be better to move logic which is currently in `httputil.NewTLSConfig` into a struct method for `TLSConfig`. In `prometheus/config/config.go`:
```
func (c *TLSConfig) GenerateConfig() (tls.Config error) {
...
}
```
If that was the case, I could simply call `module.TCP.TLSConfig.GenerateConfig()` in blackbox_exporter.

**Comment by brian-brazil** [comment]
The idea would be to take all that code that's currently in the prometheus repo and break it out to the common repo. Similar has to be done for #23 for non-ssl auth. Reaching into the internals of other repos is unclean.

**Comment by hudashot** [comment]
I've added prometheus/common#31 to common and made this PR use that code.

**Comment by hudashot** [comment]
@brian-brazil, so what do you think about this?

File: README.md [hunk 2]
```diff
@@ -39,6 +39,14 @@ modules:
...
     prober: tcp
     timeout: 5s
+  pop3s_banner:
+    prober: tcp
+    tcp:
+      query_response:
+      - expect: "^+OK"
+      tls: true
+      tls_config:
+        insecure_skip_verify: true
```
Comment (brian-brazil) [inline README.md]: Our examples should encourage good security, so let's leave the cert verification enabled by default

File: README.md [hunk 3]
```diff
@@ -39,6 +39,14 @@ modules:
   tcp_connect:
     prober: tcp
     timeout: 5s
+  pop3s_banner:
```
Comment (brian-brazil) [inline README.md]: This should be added to blackbox.yml

**Comment by knyar** [comment]
> Our examples should encourage good security, so let's leave the cert verification enabled by default

What would be the right way to document that option? Should I add two more sections into `blackbox.yml`: one with `tls: true` and the second one with `insecure_skip_verify` additionally set?

**Comment by brian-brazil** [comment]
Probably easiest to have the example with `insecure_skip_verify: false`. We'll have to figure out how we're going to document these common options (alertmanager will probably end up needing them too).

**Comment by brian-brazil** [comment]
Please squash the commits too when you're ready.

**Comment by knyar** [comment]
Updated `blackbox.yml`, squashed the commits.

**Comment by hudashot** [comment]
Updated `blackbox.yml`, squashed the commits.

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**